File: LeetCode/two_sum.go
// // Brute force O(n^2), O(1)
// func twoSum(nums []int, target int) []int {
//     ptr1 := 0
//     ptr2 := 1
//     nums_len := len(nums)
    
//     for ptr2 < nums_len {
//         if nums[ptr1] + nums[ptr2] == target {
//             break
//         }
        
//         if ptr2 + 1 == nums_len {
//             ptr1 += 1
//             ptr2 = ptr1
//         }
        
//         ptr2 += 1
//     }
    
//     return []int{ptr1, ptr2}
// }

// O(n), O(n)
func twoSum(nums []int, target int) []int {
    complement_map := make(map[int]int)
    for i, num := range nums {
        if j, exist := complement_map[num]; exist {
            return []int{i, j}
        } else {
            complement_map[target - num] = i
        }
    }
    
    return []int{}
}

File: LeetCode/two_sum.py
class Solution:
#     # Brute force, O(n^2), O(1)
#     def twoSum(self, nums: List[int], target: int) -> List[int]:
#         len_nums = len(nums)
#         for i in range(len_nums):
#             for j in range(i + 1, len_nums):
#                 if nums[i] + nums[j] == target:
#                     return [i, j]
#         raise Exception('No two sum solution')

    # O(n), O(n)
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        nums_hash = dict()
        for i, n in enumerate(nums):
            complement = target - n
            if complement in nums_hash:
                return [nums_hash[complement], i]
            nums_hash[n] = i
        raise Exception('No two sum solution')